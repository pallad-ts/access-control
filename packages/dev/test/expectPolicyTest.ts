import '@src/index';
import {ACCESS_QUERY, ACTION} from "./fixtures";
import * as sinon from 'sinon';

describe('expectPolicy', () => {
	describe('policyToAllow', () => {
		it('fails if provided value is not a function', () => {
			return expect(
				expect({})
					.policyToAllow(ACCESS_QUERY)
			)
				.rejects
				.toThrowErrorMatchingSnapshot();
		});

		it('fails if provided policy denies access for given query', async () => {
			const policy = sinon.stub().returns(false);
			await expect(
				expect(policy)
					.policyToAllow(ACCESS_QUERY)
			)
				.rejects
				.toThrowErrorMatchingSnapshot();
			sinon.assert.calledWith(policy, ACCESS_QUERY);
		});

		it('fails if provided policy abstain from voting', async () => {
			const policy = sinon.stub().returns(undefined);
			await expect(
				expect(policy)
					.policyToAllow(ACCESS_QUERY)
			)
				.rejects
				.toThrowErrorMatchingSnapshot();
			sinon.assert.calledWith(policy, ACCESS_QUERY);
		});

		it('success if provided policy allows access for given query', async () => {
			const policy = sinon.stub().returns(true);
			await expect(policy)
				.policyToAllow(ACCESS_QUERY);
			sinon.assert.calledWith(policy, ACCESS_QUERY);
		});
	});

	describe('policyToDeny', () => {
		it('fails if provided value is not a function', () => {
			return expect(
				expect({})
					.policyToDeny(ACCESS_QUERY)
			)
				.rejects
				.toThrowErrorMatchingSnapshot();
		});

		it('fails if provided policy abstain from voting', async () => {
			const policy = sinon.stub().returns(undefined);
			await expect(
				expect(policy)
					.policyToDeny(ACCESS_QUERY)
			)
				.rejects
				.toThrowErrorMatchingSnapshot();
			sinon.assert.calledWith(policy, ACCESS_QUERY);
		});

		it('fails if provided policy allows access for given query', async () => {
			const policy = sinon.stub().returns(true);
			const result = expect(policy)
				.policyToDeny(ACCESS_QUERY);
			await expect(
				result
			)
				.rejects
				.toThrowErrorMatchingSnapshot();
			sinon.assert.calledWith(policy, ACCESS_QUERY);
		});

		it('success if provided policy denies access for given query', async () => {
			const policy = sinon.stub().returns(false);
			await expect(policy)
				.policyToDeny(ACCESS_QUERY);
			sinon.assert.calledWith(policy, ACCESS_QUERY);
		});
	});
});
