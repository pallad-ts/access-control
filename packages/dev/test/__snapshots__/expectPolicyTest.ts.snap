// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expectPolicy policyToAllow fails if provided policy abstain from voting 1`] = `
"Policy has abstained from voting for following access query:
{"action": "foo_action", "principal": {"some": "principal"}, "subject": {"some": "subject"}}"
`;

exports[`expectPolicy policyToAllow fails if provided policy denies access for given query 1`] = `
"expected to ALLOW for following combination:
{"action": "foo_action", "principal": {"some": "principal"}, "subject": {"some": "subject"}}"
`;

exports[`expectPolicy policyToAllow fails if provided value is not a function 1`] = `"Policy must be a function"`;

exports[`expectPolicy policyToDeny fails if provided policy abstain from voting 1`] = `
"Policy has abstained from voting for following access query:
{"action": "foo_action", "principal": {"some": "principal"}, "subject": {"some": "subject"}}"
`;

exports[`expectPolicy policyToDeny fails if provided policy allows access for given query 1`] = `
"expected to DENY for following combination:
{"action": "foo_action", "principal": {"some": "principal"}, "subject": {"some": "subject"}}"
`;

exports[`expectPolicy policyToDeny fails if provided value is not a function 1`] = `"Policy must be a function"`;
